src/charm.py

```diff
@@ -120,11 +132,16 @@ def _on_config_changed_or_upgrade(self, event):
             # needs to intervene manually
             return
         r = self.framework.model.relations.get('storage-backend')[0]
+        ctx = CinderThreeParContext(charm_config, svc_name)
+        if not isinstance(ctx, dict):
+            # Invalid config option.
+            self.unit.status = ctx
+            return
```

Comment:
This is exactly the issue with not raising an exception, and using a special return value to indicate an error, and thus *if you forget to do this* then it would break unexpectedly.  Not handling an exception causes a program failure in a much more predictable manner, as there is no possibility that the code will continue down a path with invalid data.